pkg/assembler/backends/neo4j/ingest_pkg.go

```diff
@@ -0,0 +1,396 @@
+	if err != nil {
+		return err
+	}
+	// pkg:pypi/django@1.11.1
+	err = client.registerPackage("pypi", "", "django", "1.11.1", "")
+	if err != nil {
+		return err
+	}
+	// pkg:pypi/django@1.11.1#subpath
+	err = client.registerPackage("pypi", "", "django", "1.11.1", "subpath")
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *neo4jClient) registerPackage(packageType, namespace, name, version, subpath string, qualifiers ...string) error {
+	collectedPkg := pkgNode{}
+	collectedType := pkgType{pkgType: packageType}
+	collectedNamespace := pkgNamespace{namespace: namespace}
+	collectedName := pkgName{name: name}
+	collectedVersion := pkgVersion{version: version, subpath: subpath, qualifier: map[string]interface{}{}}
+	for _, kv := range qualifiers {
+		pair := strings.Split(kv, "=")
+		collectedVersion.qualifier[pair[0]] = pair[1]
+	}
+	pkgToTypeEdge := pkgToType{collectedPkg, collectedType}
+	typetoNamespaceEdge := typeToNamespace{collectedType, collectedNamespace}
+	namespaceToNameEdge := namespaceToName{collectedNamespace, collectedName}
+	nameToVersionEdge := nameToVersion{collectedName, collectedVersion}
+	assemblerinput := assembler.AssemblerInput{
+		Nodes: []assembler.GuacNode{collectedPkg, collectedType, collectedNamespace, collectedName, collectedVersion},
+		Edges: []assembler.GuacEdge{pkgToTypeEdge, typetoNamespaceEdge, namespaceToNameEdge, nameToVersionEdge},
+	}
+	err := assembler.StoreGraph(assemblerinput, c.driver)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// pkgNode represents the top level pkg->Type->Namespace->Name->Version
+type pkgNode struct {
+}
+
+func (pn pkgNode) Type() string {
+	return "Pkg"
+}
+
+func (pn pkgNode) Properties() map[string]interface{} {
+	properties := make(map[string]interface{})
+	properties["pkg"] = "pkg"
+	return properties
+}
+
+func (pn pkgNode) PropertyNames() []string {
+	fields := []string{"pkg"}
+	return fields
+}
+
+func (pn pkgNode) IdentifiablePropertyNames() []string {
+	return []string{"pkg"}
+}
+
+type pkgType struct {
+	pkgType string
+}
+
+func (pt pkgType) Type() string {
+	return "PkgType"
+}
+
+func (pt pkgType) Properties() map[string]interface{} {
+	properties := make(map[string]interface{})
+	properties["type"] = pt.pkgType
+	return properties
+}
+
+func (pt pkgType) PropertyNames() []string {
+	fields := []string{"type"}
+	return fields
+}
+
+func (pt pkgType) IdentifiablePropertyNames() []string {
+	return []string{"type"}
+}
+
+type pkgNamespace struct {
+	namespace string
+}
+
+func (pn pkgNamespace) Type() string {
+	return "PkgNamespace"
+}
+
+func (pn pkgNamespace) Properties() map[string]interface{} {
+	properties := make(map[string]interface{})
+	properties["namespace"] = pn.namespace
+	return properties
+}
+
+func (pn pkgNamespace) PropertyNames() []string {
+	fields := []string{"namespace"}
+	return fields
+}
+
+func (pn pkgNamespace) IdentifiablePropertyNames() []string {
+	return []string{"namespace"}
```

Comment:
i just realized this could be complicated... the issue is that trie lookup assumes that uniqueness of the lookup is based on the sequence of traversals that form the prefix, the current way to storegraph would not allow this. Not looking for an answer for this PR, but just wondering what you're thinking about with the neo4j schema for that case.

Reply:
Let's discuss this further.